### ISSUE TITLE
OOM when redirecting large stdout to asyncio Process

### ISSUE THREAD

**Issue by BenjyWiener** [ronf/asyncssh#664]
When redirecting stdout to an asyncio `StreamWriter`, all data is written immediately to the writer, without any calls to `drain`. This becomes a problem when piping to a local subprocess that reads slower than data is transferred.

Example: reading a tar file on a remote machine and piping it to tar on the local machine:

```python
tar_subproc = await asyncio.subprocess.create_subprocess_exec(
    'tar', '-xzf', '-',
    stdin=asyncio.subprocess.PIPE,
)
async with asyncio.connect(...) as conn:
    await conn.run('cat verybig.tar.gz', stdout=tar_subproc.stdin)
```

A workaround is to use `create_subprocess`, read from stdout in a loop, write to `tar_subproc.stdin` and await `tar_subproc.stdin.drain()` in the loop. But this feels clunky compared to the simplicity of using `redirect`.

**Comment by ronf** [comment]
Thanks for the report.

In theory, there should be flow control here. The expectation is that the pipe transport used to write to the local process should be calling its associated protocol's `pause_writing()` method when the write buffer exceeds a certain size. In this case, the protocol object is an AsyncSSH `PipeWriter`, and that should trigger a call to `pause_feeding()` on the AsyncSSH process. That in turn can cause the process to pause reading on the channel if the receive buffer gets too large. The default is to allow only up to the receive window size, which defaults to 2 MBytes.

I'll see if I can reproduce this over the weekend, and get back to you once I know more.

**Comment by BenjyWiener** [comment]
Looking at the source of `_create_writer`, if `target` is an `asyncio.StreamWriter`, it just gets wrapped in a `_StreamWriter`, which doesn't hold any reference to the `SSHProcess`, so it's not actually using the pipe protocol.

**Comment by BenjyWiener** [comment]
For reproducing, here's a small script that writes data faster than it's read:

```

import asyncio
import asyncssh


async def main():
  loc_proc = await asyncio.subprocess.create_subprocess_shell(
    'while head -c 1; do sleep 1; done'
  )

  async with asyncssh.connect('localhost') as conn:
    await conn.run(
      'dd if=/dev/zero bs=1M count=5000'
    )

  await loc_proc.wait()


asyncio.run(main())
```

**Comment by ronf** [comment]
Thanks for the sample code!

Looking more closely at that, I see you're using the asyncio version of the subprocess creation code, and that's why the code ends up using _StreamWriter rather than _PipeWriter as I originally thought. Looking at _StreamWriter, you're right that it doesn't do any flow control on writing at the moment. That's something I can add, but it may involve some extra overhead of having a dedicated writing task, and feeding data to it through a queue.

As a workaround for now, you can actually use a callback version with proper flow control if you just change from `asyncio.subprocess.create_subprocess_shell()` to `subprocess.Popen()`. For instance:

```python
import asyncio
import asyncssh
import subprocess

async def main():
  loc_proc = subprocess.Popen(
    'while head -c 1; do sleep 1; done',
    shell=True, stdin=subprocess.PIPE,
  )

  async with asyncssh.connect('localhost') as conn:
    await conn.run(
      'dd if=/dev/zero bs=1M count=5000',
      stdout=loc_proc.stdin,
    )

asyncio.run(main())
```

This uses _PipeWriter, which receives callbacks when the destination buffer is full. Since it is all callback-based, it's actually cheaper than trying to do this with an asyncio stream as a redirect target.

**Comment by BenjyWiener** [comment]
I think that solves my problem, thanks!

After doing a bit more research it seems that the `asyncio.subprocess.create_subprocess_*` functions aren't really meant for piping data between processes. (What I was trying to do wouldn't work even between two `asyncio` subprocess.)

**Comment by ronf** [comment]
Yeah - I don't think you can pass an already existing open stream (an asyncio StreamReader or StreamWriter) as stdin/stdout/stderr to asyncio.subprocess. Support for a much wider range of argument types is specific to AsyncSSH's version of redirection.

That said, it would be good if flow control worked properly in cases where someone did pass an asyncio.StreamWriter into AsyncSSH, whether from asyncio.subprocess or any other existing stream writers an application might create. I think I have this working in commit f2020ed in the "develop" branch if you'd like to take a look...

**Comment by BenjyWiener** [comment]
Thanks! I just tested with the latest commit, but I'm still seeing a memory leak.

I think now the queue is filling up instead of the write buffer, but the effect is the same.

I think we might need to pause reading before calling `drain()` and resume immediately after.

Alternatively, we can manually check if `writer.transport. get_write_buffer_size() > writer.transport. get_write_buffer_limits()`, and only then pause reading while `await`ing `writer.drain()`.

**Comment by ronf** [comment]
Thanks - you're absolutely right. I adapted this new code from the aiofile case, but that has the same problem. It doesn't actually trigger the back pressure in the process, so the queue can grow without bound. I'll need to think a bit about how I want to handle this in both of these places. For the aiofile case in particular, I'm not sure there's any good way to determine in advance whether a `write()` to the file will block before actually making the call. I could potentially do something based on queue length there, but I'll think about it a bit.

**Comment by ronf** [comment]
Ok - I've added proper back pressure for both AsyncFilewriter and StreamWriter into the "develop" branch in commit 24b60d1. Some additional memory is still required because a pause feeding from the channel doesn't really take effect until the window gets full and that defaults to 2 MiB of data, but this commit prevents the queue from growing without bound.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.15.0.

**Comment by BenjyWiener** [comment]
Amazing, thank you!
